admin/acceptance.py

```diff
@@ -528,19 +577,196 @@ def start_cluster(self, reactor):
+            return failure
+
+        d.addErrback(error)
+        return d
+
+    def _create_nodes(self, reactor, names):
+        """
+        Create and provision a number of nodes with the given names.
+
+        :param reactor: The reactor.
+        :param names: The names.
+        :type names: list of str
+        :return: List of Deferreds each firing when the corresponding node
+            is created and provisioned.
+        """
+        reactor.suggestThreadPoolSize(len(names))
+        return [
+            self._create_and_provision(reactor, name) for name in names
+        ]
+
+    def _make_node_name(self, tag, index):
+        """
+        Construct a node name out of various components.
+        """
+        return "%s-%s-%s-%d" % (
+            self.identity.prefix, self.creator, tag, index,
+        )
+
+    def _add_node_to_cluster(self, reactor, cluster, node, index):
+        """
+        Configure the given node as an agent node for the given cluster.
+
+        :param reactor: The reactor.
+        :param reactor: The cluster.
+        :param Node node: The node.
+        :param int index: The index of the ndoe in the cluster.
+        :return: Deferred that fires with the :param:`node` when it is
+            configured.
+        """
+        node_certnkey = cluster.certificates.add_node(index)
+        commands = configure_node(
+            cluster,
+            node,
+            node_certnkey,
+            self.dataset_backend_configuration,
+            'libcloud',
+            logging_config=self.config.get('logging'),
+        )
+        d = perform(make_dispatcher(reactor), commands)
+
+        def add_node(ignored):
+            if node is not cluster.control_node:
+                cluster.all_nodes.append(node)
+            cluster.agent_nodes.append(node)
+            return node
+
+        def error(failure):
+            print "Failed to configure node {}".format(node.name)
+            write_failure(failure)
+            if node is not cluster.control_node:
+                self._destroy_node(node)
+            # Failure to add a node to a cluster is not fatal.
+            # As long as we have the control node we can claim
+            # the partial success.
+            return failure
+
+        d.addErrback(error)
+        d.addCallback(add_node)
+        return d
+
+    def extend_cluster(self, reactor, cluster, count, tag, starting_index):
+        """
+        Extend a cluster with nodes provisioned via libcloud.
+        """
+        names = []
+        for index in range(starting_index, starting_index + count):
+            names.append(self._make_node_name(tag, index))
+        results = self._create_nodes(reactor, names)
+
+        def add_node(node, index):
+            return self._add_node_to_cluster(reactor, cluster, node, index)
+
+        for i, d in enumerate(results):
+            d.addCallback(add_node, starting_index + i)
+
+        return results
+
+    @inlineCallbacks
```

Comment:
I can see `yield`s, but not in the added code, later in the function.

inlineCallbacks do get used occasionally, always outside the core Flocker code, and more relevantly, in the acceptance testing code from which this code was derived. You can consider it the "grandfathering" loophole of the engineering guide.